crates/accelerate/src/synthesis/linear/utils.rs

```diff
@@ -0,0 +1,200 @@
+                    new_k = j;
+                    new_r = i;
+                    break;
+                }
+            }
+            if is_non_zero {
+                break;
+            }
+        }
+        if !is_non_zero {
+            return perm; // A is in the canonical form
+        }
+
+        if new_r != r {
+            let temp_r = mat.slice_mut(s![r, ..]).to_owned();
+            let temp_new_r = mat.slice_mut(s![new_r, ..]).to_owned();
+            mat.slice_mut(s![r, ..]).assign(&temp_new_r);
+            mat.slice_mut(s![new_r, ..]).assign(&temp_r);
+            perm.swap(r, new_r);
+        }
+
+        // Copy source row to avoid trying multiple borrows at once
+        let row0 = mat.row(r).to_owned();
+        mat.axis_iter_mut(Axis(0))
+            .enumerate()
+            .filter(|(i, row)| {
+                (full_elim == Some(true) && (*i < r) && row[new_k])
+                    || (*i > r && *i < m && row[new_k])
+            })
+            .for_each(|(_i, mut row)| {
+                row.zip_mut_with(&row0, |x, &y| *x ^= y);
+            });
+
+        r += 1;
+    }
+    perm
+}
+
+/// Given a boolean matrix A after Gaussian elimination, computes its rank
+/// (i.e. simply the number of nonzero rows)
+pub fn compute_rank_after_gauss_elim_inner(mat: ArrayView2<bool>) -> usize {
+    let rank: usize = mat
+        .axis_iter(Axis(0))
+        .map(|row| row.fold(false, |out, val| out | *val) as usize)
+        .sum();
+    rank
+}
+
+/// Given a boolean matrix mat computes its rank
+pub fn compute_rank_inner(mat: ArrayView2<bool>) -> usize {
+    let mut temp_mat = mat.to_owned();
+    gauss_elimination_with_perm_inner(temp_mat.view_mut(), None, Some(false));
+    let rank = compute_rank_after_gauss_elim_inner(temp_mat.view());
+    rank
+}
+
+/// Given a square boolean matrix mat, tries to compute its inverse.
+pub fn calc_inverse_matrix_inner(
+    mat: ArrayView2<bool>,
+    verify: bool,
+) -> Result<Array2<bool>, String> {
+    if mat.shape()[0] != mat.shape()[1] {
+        return Err("Matrix to invert is a non-square matrix.".to_string());
+    }
+    let n = mat.shape()[0];
+
+    // concatenate the matrix and identity
+    let identity_matrix: Array2<bool> = Array2::from_shape_fn((n, n), |(i, j)| i == j);
+    let mut mat1 = concatenate(Axis(1), &[mat.view(), identity_matrix.view()]).unwrap();
+
+    gauss_elimination_with_perm_inner(mat1.view_mut(), None, Some(true));
+
+    let r = compute_rank_after_gauss_elim_inner(mat1.slice(s![.., 0..n]));
+    if r < n {
+        return Err("The matrix is not invertible.".to_string());
+    }
+
+    let invmat = mat1.slice(s![.., n..2 * n]).to_owned();
+
+    if verify {
+        let mat2 = binary_matmul_inner(mat, (&invmat).into())?;
+        let identity_matrix: Array2<bool> = Array2::from_shape_fn((n, n), |(i, j)| i == j);
+        if mat2.ne(&identity_matrix) {
+            return Err("The inverse matrix is not correct.".to_string());
+        }
+    }
```

Comment:
There is a parameter "verify" in several parts of the synthesis code, which is mainly used for tests, since the algorithms are quite complicated (the default value is False)

Reply:
We can leave it for this PR 👍🏻 but a common subroutine like matrix inversion seems like something stable enough that we don't need to bake in a verification and instead just have some tests in the test suite (fwiw other packages like NumPy/SciPy also don't have any arguments to check that 🙂).